### DISCUSSION TITLE
Interactive PyVista inside a Jupyter Notebook

### DISCUSSION THREAD

**Discussion by GuillaumeFavelier** [pyvista/pyvista#2606] [Q&A]
I'm currently experimenting with `PyVista` inside a `jupyter notebook` and I noticed that the figures required a specific pipeline to be displayed automatically.

Do you know a way to see the result of the first scenario `plot_scene_1`?

![image](https://user-images.githubusercontent.com/18143289/59131001-501e8d00-8971-11e9-8bd2-a7a89914a6c2.png)

**Comment by GuillaumeFavelier** [comment]
Actually if I modify `plot_scene_1()` like the following, it is displayed automatically:

```py
def plot_scene_1():
    S = pv.Sphere()
    plotter = pv.Plotter()
    plotter.add_mesh(S)
    return plotter.show()
plot_scene_1()
```
So I work on integrating this idea in [mne-python](https://github.com/mne-tools/mne-python/pull/6437). Please don't hesitate to tell me if you know any way to do it better.

**Comment by banesullivan** [comment]
@GuillaumeFavelier: the `plotter.show()` call is actually returning a `panel.pane.vtk.VTK` object that can be displayed in the Jupyter cell output. Calling `plotter.show()` and not having its result be the cell's output will result in the panel-based interactive rendering not be shown.

<img width="929" alt="Screen Shot 2019-06-08 at 5 53 29 PM" src="https://user-images.githubusercontent.com/22067021/59152649-e4a7ee80-8a48-11e9-82b9-16893ac2a99e.png">


I expect this behavior... however this might be considered a bug as a user might expect that any call to `.show()` will show the renderer. Also, this is inconsistent with a the static screenshot display in Jupyter notebooks (which returns None and displays the screenshot without it being the cells output):

```py
from pyvista import examples
import pyvista as pv

mesh = examples.download_st_helens().warp_by_scalar()

plotter = pv.Plotter()
plotter.add_mesh(mesh)
output = plotter.show(use_panel=False)
```

Perhaps we need to collaborate with the `panel` developers to see if there is a way to display a panel object without it being a cell's output?

**Comment by GuillaumeFavelier** [comment]
Many thanks for the detailed explanations @banesullivan, now I understand better how it works and I think I can manage with this solution for now. Although I had to rely on the part of code detecting if the script is executed inside a notebook.

> Perhaps we need to collaborate with the panel developers to see if there is a way to display a panel object without it being a cell's output?

That would wonderful of course!

**Comment by banesullivan** [comment]
> Although I had to rely on the part of code detecting if the script is executed inside a notebook.

That's strange, what exactly do you mean? I thought I had the `panel` code to only run if the code is being executed from a Jupyter notebook.

**Comment by GuillaumeFavelier** [comment]
> I thought I had the panel code to only run if the code is being executed from a Jupyter notebook.

And yes, it works as expected. My bad, it was not clear. What I mean is, I had to use a workaround in my specific use case because I expect that any call to `.show()` will show the renderer (*c.f.* `plot_sphere_1()` scenario) even if it's not the cell's output.

**Comment by GuillaumeFavelier** [comment]
Actually after some more research I found a way to force panel to show the pane thanks to the `show_server()` function:

![image](https://user-images.githubusercontent.com/18143289/61162684-fda13500-a509-11e9-8e9e-ecda31d57ca6.png)

So I used the following function:

```py
def notebook_show(disp, notebook_url="localhost:8888", port=0):
    from panel.viewable import show_server
    show_server(disp, notebook_url, port)
```

Do you think there would be a way to integrate such a feature in PyVista @banesullivan ? Eventually, I can open a discussion with the Panel developers to find a better/cleaner solution just in case.

**Comment by banesullivan** [comment]
> Do you think there would be a way to integrate such a feature in PyVista

Probably - we should add something like this in https://github.com/pyvista/pyvista/pull/300. My only hesitation is that the url and port have to be known, and they might not always be `notebook_url="localhost:8888"`, `port=0` so maybe we should ping the `panel` dev team.

Frankly, I think all notebook related dev should be paused until a few things move forward with `itk-jupyter-widgets`. Reference https://github.com/InsightSoftwareConsortium/itk-jupyter-widgets/pull/154 and waiting on:

- https://github.com/InsightSoftwareConsortium/itk-jupyter-widgets/issues/163
- https://github.com/InsightSoftwareConsortium/itk-jupyter-widgets/issues/166

Once those two issues are addressed, we will favor using `itk-jupyter-widgets` for notebook rendering *directly* instead of all the hacky solutions where a VTK render window is exported/serialized for VTKjs, X3D, or K3D.

**Comment by GuillaumeFavelier** [comment]
After more investigation I found that this version of `plot_scene_1()` is way simpler and doesn't require `notebook_show()` or any unnecessary tweaks anymore:
```py
def plot_scene_1():
    from IPython.display import display
    S = pv.Sphere()
    plotter = pv.Plotter()
    plotter.add_mesh(S, color='white')
    disp = plotter.show(use_panel=True, auto_close=False)
    display(disp)
plot_scene_1()
```
